### PR TITLE
[6X back-port] [ORCA] Add support for ordered and multi-distinct aggregates

### PR DESCRIPTION
**Pull request by dgkimura**
Cherry-pick of commit https://github.com/greenplum-db/gpdb/commit/ad40cc672d11b14e269a2626ba37e0b88e6b1617.

There were a few flavor of conflicts
1) C++11 features which can't be used in 6X
2) Code refactors existing only in master branch that caused conflicts
3) Aggref node diff (aggargtypes field doesn't exist in 6X branch)
4) Miscellaneous diffs

_Important Note_:  Grouping sets with ordered aggregates does not work.  For example:
```sql
select grouping(a), a, array_agg(b),
       rank(a) within group (order by b nulls first),
       rank(a) within group (order by b nulls last)
from (values (1,1),(1,4),(1,5),(3,1),(3,2)) v(a,b)
group by rollup (a) order by a;
 ```
 
 There's a GUC optimizer_enable_orderedagg from commit https://github.com/greenplum-db/gpdb/commit/90cfe6e38433de366e1f0973a07929978c3ff0de that disables this feature by default.

### CODE REVIEW

**Review by bhuvnesh2703 — APPROVED** [review]
LGTM, Minor comments, ignore if you like.

File: src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 1]
```diff
@@ -6589,12 +6593,43 @@ CTranslatorExprToDXL::PdxlnScAggref(CExpression *pexprAggFunc)
 		edxlaggrefstage = EdxlaggstagePartial;
 	}
 
-	CDXLScalarAggref *pdxlopAggRef = GPOS_NEW(m_mp)
-		CDXLScalarAggref(m_mp, pmdidAggFunc, resolved_rettype,
-						 popScAggFunc->IsDistinct(), edxlaggrefstage);
+	EdxlAggrefKind edxlaggrefkind = EdxlaggkindNormal;
```
[the remark this answers is not in the document]
Comment (dgkimura) [reply src/backend/gporca/libgpopt/src/translate/CTranslatorExprToDXL.cpp:6596]: Good point.  I'll add default case and throw.  Currently `pg_aggregate.h` defines 3 cases, but throwing will let us know when that changes and protect against bad input.

File: src/backend/utils/misc/guc_gp.c [hunk 2]
```diff
@@ -3053,7 +3054,7 @@ struct config_bool ConfigureNamesBool_gp[] =
 
 	{
 		{"optimizer_enable_eageragg", PGC_USERSET, DEVELOPER_OPTIONS,
-			gettext_noop("Enable Eager Agg transform for pushing aggregate below an innerjoin."),
+			gettext_noop("Enable ordered aggregate plans."),
```
[the remark this answers is not in the document]
Comment (dgkimura) [reply src/backend/utils/misc/guc_gp.c:3057]: Wow.. thanks for catching that.

File: src/backend/gporca/libgpopt/src/operators/CScalarSortGroupClause.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+CScalarSortGroupClause::OsPrint(IOstream &os) const
+{
+	os << SzId();
+	os << "(tleSortGroupRef:" << m_tle_sort_group_ref << ",";
+	os << "eqop:" << m_eqop << ",";
+	os << "sortop:" << m_sortop << ",";
+	os << "nulls_first:" << (m_nulls_first ? "true" : "false") << ",";
+	os << "hashable:" << (m_hashable ? "true" : "false") << ")";
+	return os;
+}
```
[the remark this answers is not in the document]
Comment (dgkimura) [reply src/backend/gporca/libgpopt/src/operators/CScalarSortGroupClause.cpp:81]: Yeah, thanks. Fixed by added DbgPrintMixin to CScalarSortGroupClause interface and validating that it prints out correctly.

File: src/backend/gporca/libgpopt/src/translate/CTranslatorDXLToExpr.cpp [hunk 4]
```diff
@@ -3056,6 +3059,26 @@ CTranslatorDXLToExpr::PexprAggFunc(const CDXLNode *pdxlnAggref)
 	}
 	BOOL fSplit = (EdxlaggstageNormal != dxl_op->GetDXLAggStage());
 
+	EAggfuncKind agg_func_kind = EaggfunckindNormal;
```
[the remark this answers is not in the document]
Comment (dgkimura) [reply src/backend/gporca/libgpopt/src/translate/CTranslatorDXLToExpr.cpp:3062]: Fixed by raising if doesn't match one of expected cases.

File: src/backend/gporca/libgpopt/src/xforms/CXformEagerAgg.cpp [hunk 5]
```diff
@@ -352,15 +354,32 @@ CXformEagerAgg::PopulateUpperProjectElement(
...
+						   EaggfuncstageGlobal, true, NULL, EaggfunckindNormal);
 
 	// populate the argument list for the upper aggregate function
 	CExpressionArray *upper_agg_arg_array = GPOS_NEW(mp) CExpressionArray(mp);
-	upper_agg_arg_array->Append(CUtils::PexprScalarIdent(mp, lower_colref));
+
+	CExpressionArray *args = GPOS_NEW(mp) CExpressionArray(mp);
+	args->Append(CUtils::PexprScalarIdent(mp, lower_colref));
+
+	upper_agg_arg_array->Append(
```
[the remark this answers is not in the document]
Comment (dgkimura) [reply src/backend/gporca/libgpopt/src/xforms/CXformEagerAgg.cpp]: Fixed up in commit https://github.com/greenplum-db/gpdb/pull/12992/commits/5297535748917af8cd176a3e12085e471fa046b7 via `CUtils::PexprAggFuncArgs()`.  Looks cleaner. Thanks!

File: src/backend/gporca/libgpopt/src/base/CUtils.cpp [hunk 6]
```diff
@@ -1716,13 +1716,31 @@ CUtils::PexprAggFunc(CMemoryPool *mp, IMDId *pmdidAggFunc,
...
 	// generate function arguments
-	CExpression *pexprScalarIdent = PexprScalarIdent(mp, colref);
 	CExpressionArray *pdrgpexpr = GPOS_NEW(mp) CExpressionArray(mp);
-	pdrgpexpr->Append(pexprScalarIdent);
+
+	CExpression *pexprScalarIdent = PexprScalarIdent(mp, colref);
+	CExpressionArray *pdrgpexprArgs = GPOS_NEW(mp) CExpressionArray(mp);
+	pdrgpexprArgs->Append(pexprScalarIdent);
+
+	pdrgpexpr->Append(GPOS_NEW(mp) CExpression(
```
[the remark this answers is not in the document]
Comment (dgkimura) [reply src/backend/gporca/libgpopt/src/base/CUtils.cpp:1737]: Added  `CUtils::PexprAggFuncArgs()`

**Pull request merged**